File: 146-lru-cache.py
'''
146. LRU Cache

Medium

Design a data structure that follows the constraints of a Least Recently Used (LRU) cache.

Implement the LRUCache class:

LRUCache(int capacity) Initialize the LRU cache with positive size capacity.

int get(int key) Return the value of the key if the key exists, otherwise return -1.

void put(int key, int value) Update the value of the key if the key exists. 
Otherwise, add the key-value pair to the cache. 
If the number of keys exceeds the capacity from this operation, evict the least recently used key.

The functions get and put must each run in O(1) average time complexity.

Constraints:
1 <= capacity <= 3000
0 <= key <= 104
0 <= value <= 105
At most 2 * 105 calls will be made to get and put.

Notes:
(1) Create a linked list that deletes from head and adds at tail. 
This allows O(1) time complexity for both operations.
(2) Use a dictionary to map keys to nodes in the linked list for O(1) access.
(3) When a key is accessed, remove it from its current position and reinsert it at the tail of the linked list to mark it as recently used.
(4) When the cache exceeds its capacity, remove the node at the head of the linked list, which represents the least recently used key.  
'''

class Node:
    def __init__(self, k=0, v=0):
        self.key = k
        self.val = v
        self.prev = None
        self.next = None 

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = Node()
        self.tail = Node()
        self.cache = {}
        self.head.next = self.tail
        self.tail.prev = self.head
        

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self._remove(node)
            self._insert(node)
            return node.val
        return -1 
        
    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self._remove(self.cache[key])
        node = Node(key, value)
        self._insert(node)
        self.cache[key] = node
        while len(self.cache) > self.capacity:
            tmp = self.head.next
            self._remove(tmp)
            del self.cache[tmp.key]
    def _remove(self, node):
        prev = node.prev
        nxt = node.next
        prev.next = nxt
        nxt.prev = prev
    
    def _insert(self, node):
        prev = self.tail.prev
        prev.next = node
        self.tail.prev = node
        node.prev = prev
        node.next = self.tail


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache.java
import java.util.HashMap;

class Node {
    int key;
    int val;
    Node prev;
    Node next;

    public Node(int key, int val) {
        this.key = key;
        this.val = val;
    }

}

class LRUCache {

    Map<Integer, Node> cache = new HashMap<>();
    int capacity;
    Node head;
    Node tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.head = new Node(0,0);
        this.tail = new Node(0,0);

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    public void remove(Node node) {
        Node prev = node.prev;
        Node next = node.next;
        prev.next = next;
        next.prev = prev;
    }

    public void insert(Node node) {
        Node prev = this.tail.prev;
        Node next = this.tail;
        prev.next = node;
        next.prev = node;
        node.prev = prev;
        node.next = next;
    }

    public int get(int key) {
        if (this.cache.containsKey(key)) {
            Node tmp = this.cache.get(key);
            remove(tmp);
            insert(tmp);
            return tmp.val;
        } else {
            return -1;
        }
    }

    public void put(int key, int val) {
        if (this.cache.containsKey(key)) {
            Node tmp = this.cache.get(key);
            remove(tmp);
            insert(tmp);
            tmp.val = val;
        } else {
            Node tmp = new Node(key, val);
            insert(tmp);
            this.cache.put(key, tmp);
        }

        while (this.cache.size() > this.capacity) {
            Node tmp = this.head.next;
            remove(tmp);
            this.cache.remove(tmp.key);
        }
    }
}



/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */